### ISSUE TITLE
[optimization, quantization, inference] Clarification regarding docs/API

### ISSUE THREAD

**Issue by xor-xor** [microsoft/onnxruntime#6851]
Hi,

I have a BERT model exported from PyTorch as ONNX and then optimized & quantized with `onnxruntime_tools`. All of this is performed in offline manner, on the same machine. Next, the resulting model is served for predictions with `onnxruntime` and `CPUExecutionProvider`, on different machines. Here are the relevant code snippets:

- export to ONNX, optimization and quantization (same machine):
```python
import torch
from onnxruntime.quantization import QuantType, quantize_dynamic
from onnxruntime_tools import optimizer
from onnxruntime_tools.transformers.onnx_model_bert import BertOptimizationOptions

# loading weights into `model` variable omitted for brevity

with torch.no_grad():
    inputs = {
        "input_ids": torch.ones(1, 128, dtype=torch.int64),
        "attention_mask": torch.ones(1, 128, dtype=torch.int64),
        "token_type_ids": torch.ones(1, 128, dtype=torch.int64),
    }
    outputs = model(**inputs)
symbolic_names = {0: "batch_size", 1: "max_seq_len"}
torch.onnx.export(
    model=model,
    args=(inputs["input_ids"], inputs["attention_mask"], inputs["token_type_ids"]),
    f="model.onnx",
    opset_version=12,
    do_constant_folding=True,
    input_names=["input_ids", "attention_mask", "token_type_ids"],
    output_names=["output"],
    dynamic_axes={
        "input_ids": symbolic_names,
        "attention_mask": symbolic_names,
        "token_type_ids": symbolic_names,
    },
)

opt_options = BertOptimizationOptions("bert")
opt_options.enable_embed_layer_norm = False
opt_model = optimizer.optimize_model(
    input="model.onnx",
    model_type="bert",
    num_heads=12,
    hidden_size=768,
    optimization_options=opt_options,
)
opt_model.save_model_to_file("model_opt")

quantize_dynamic(
    model_input="model_opt",
    model_output="model_opt_quant",
    weight_type=QuantType.QInt8,
)
```

- serving for predictions (different machines):
```python
from onnxruntime import GraphOptimizationLevel, InferenceSession, SessionOptions

options = SessionOptions()
options.graph_optimization_level = GraphOptimizationLevel.ORT_ENABLE_ALL
ort_session = InferenceSession(
	path_or_bytes="model_opt_quant",
	sess_options=options,
	providers=["CPUExecutionProvider"],
)
```

- versions of packages being used:
```
torch==1.6.0
onnxruntime==1.6.0
onnxruntime-tools==1.6.0
```

What I've been noticing recently is that on _some_ machines (equipped only with CPUs - there are no GPUs involved at optimization/quantization/inference stages), I get _slightly_ different metrics (precision/recall/f1) - despite the fact that each one of them has the same model, the same code that serves it and that I perform evaluation with the same dataset. So the only variable here is the hardware - which makes sense, considering that [your documentation][1] states the following:

> When running in offline mode, make sure to use the exact same options (e.g., execution providers, optimization level) and hardware as the target machine that the model inference will run on (e.g., you cannot run a model pre-optimized for a GPU execution provider on a machine that is equipped only with CPU).
>
> When layout optimizations are enabled, the offline mode can only be used on compatible hardware to the environment when the offline model is saved. For example, if model has layout optimized for AVX2, the offline model would require CPUs that support AVX2.

Moreover, I'm aware of [this readme][2], which states that quantization and graph optimization may not always work together.

But there are couple of things which are still a bit unclear to me:
- Which optimization options really matter when you write "make sure to use the exact same options"? You mention execution providers and optimization level - but how about things like e.g. `BertOptimizationOptions`, which doesn't have their counterparts in `SessionOptions`?
- What is the expected outcome of having `opt_level=0` during offline optimization and `ORT_ENABLE_ALL` during inference session, with quantization between them (i.e. as in the code snippets above)? How about having `opt_level=99` and `ORT_ENABLE_ALL` instead? 
- If I don't get any errors or warnings from `quantize_dynamic`, is it OK to assume that the current combination of graph optimization and quantization settings works OK?
- If I don't get any errors or warnings at the serving stage (i.e. from `InferenceSession`), is it OK to assume that everything _should_ work fine (i.e. my metrics should be the same on every machine)?
- Finally, [the docstring for `InferenceSession`][3] mentions ORT and ONNX formats - what is the difference between them?

Thanks in advance for shedding some light on the questions above.

[1]: https://www.onnxruntime.ai/docs/resources/graph-optimizations.html#onlineoffline-mode
[2]: https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/quantization/README.md#quantization-and-graph-optimization
[3]: https://github.com/microsoft/onnxruntime/blob/d761571afcbc50be94b2ffdfcaa09cfd9e2eaf5e/onnxruntime/python/onnxruntime_inference_collection.py#L238

**Comment by tianleiwu** [comment]
> Which optimization options really matter when you write "make sure to use the exact same options"? You mention execution providers and optimization level - but how about things like e.g. BertOptimizationOptions, which doesn't have their counterparts in SessionOptions?
> What is the expected outcome of having opt_level=0 during offline optimization and ORT_ENABLE_ALL during inference session, with quantization between them (i.e. as in the code snippets above)? How about having opt_level=99 and ORT_ENABLE_ALL instead?
> If I don't get any errors or warnings from quantize_dynamic, is it OK to assume that the current combination of graph optimization and quantization settings works OK?

As long as the model is optimized for CPU and your inference is also in CPU, you can use the default opt level in offline and online tool. Different optimization level might output different graph, which will get same output but different performance for optimized (but not quantized) model. For quantized model, since quantize and de-quantize (there is precision loss in quantization) nodes might place differently given different graph, the result might be different for different optimization level.

If I don't get any errors or warnings at the serving stage (i.e. from InferenceSession), is it OK to assume that everything should work fine (i.e. my metrics should be the same on every machine)?

> For dynamic quantization, if you used batch size 1 in evaluating, the metric shall be same. If you used batch_size > 1, it depends on whether quantization is on whole input or per sample in a batch.

A potential variance is whether your machine has AVX2 or AVX-512. If you find their output are different, please fill an issue.

Finally, the docstring for InferenceSession mentions ORT and ONNX formats - what is the difference between them?

That are different file format in storage. ONNX is for portable and it can be used by others (like TensorRT), and ORT can only be used in OnnxRuntime.

**Comment by xor-xor** [comment]
After some experiments, it seems that indeed, AVX-512 (VNNI?) is the source of the aforementioned variance in my metrics while using `dynamic_quantization` (and yes, I'm evaluating my models with batch size == 1). I'll try to gather some more data on this and eventually open a new issue, as suggested.

Just one more question here regarding ONNX/ORT formats: I understand that the output of `onnxruntime_tools`
(e.g. `optimizer.optimize_model(...).save_model_to_file(...)` and `dynamic_quantization` is always ORT, not ONNX - am I correct?

**Comment by xor-xor** [comment]
As for my last question, in case someone would land here (BTW, there's a [whole new section re: ORT format](https://onnxruntime.ai/docs/reference/ort-model-format.html) in the docs now):

> I understand that the output of `onnxruntime_tools` (e.g. `optimizer.optimize_model(...).save_model_to_file(...))` and `dynamic_quantization` is always ORT, not ONNX - am I correct?

No, this will be still ONNX (Protocol Buffers), whereas ORT (FlatBuffers) needs to be chosen explicitly, as it serves different purposes (applications in more constrained environments). Also, as far as I understand the code, in the latter case, apart from the model itself, the inference session is saved as well.
